tensorflow/core/kernels/fft_ops.cc

```diff
@@ -97,9 +84,141 @@ class FFTGPUBase : public OpKernel {
+  bool IsForward() const override { return Forward; }
+  bool IsReal() const override { return _Real; }
+
+  void DoFFT(OpKernelContext* ctx, const Tensor& in, uint64* fft_shape,
+             Tensor* out) override {
+    if (!IsReal()) {
+      auto input = ((Tensor) in).flat_inner_dims<complex64, FFTRank + 1>();
+
+      // Apply the functor.
+      FFTFunctor<CPUDevice, complex64, complex64, Eigen::BothParts,
+                Forward ? Eigen::FFT_FORWARD : Eigen::FFT_REVERSE,
+                FFTRank> functor;
+      functor(ctx->eigen_device<CPUDevice>(),
+              out->flat_inner_dims<complex64, FFTRank + 1>(), input);
+    }
+    else {
+      if (IsForward()) {
+        auto input = ((Tensor) in).flat_inner_dims<float, FFTRank + 1>();
+        // Create a temporary placeholder for the full FFT.
+        Tensor temp;
+        OP_REQUIRES_OK(ctx,ctx->allocate_temp(
+          DataTypeToEnum<complex64>::v(), in.shape(), &temp
+        ));
+        auto full_fft = temp.flat_inner_dims<complex64, FFTRank + 1>();
+        // Apply the functor.
+        FFTFunctor<CPUDevice, float, complex64, Eigen::BothParts,
+                  Eigen::FFT_FORWARD, FFTRank> functor;
+        functor(ctx->eigen_device<CPUDevice>(), full_fft, input);
+        // Create zero indices for slicing.
+        Eigen::DSizes<Eigen::DenseIndex, FFTRank + 1> startIndices;
+        // Convert output to tensor and get tensor size for slicing.
+        auto output = out->flat_inner_dims<complex64, FFTRank + 1>();
+        auto sizes = output.dimensions();
+        // Slice the full FFT to get the non-negative frequency components only.
+        output.slice(startIndices, sizes) =
+          full_fft.slice(startIndices, sizes);
+      }
+      else {
+        auto input = ((Tensor) in).flat_inner_dims<complex64, FFTRank + 1>();
+        // The first dimension contains the zero-frequency component which we
+        // do not want to duplicate. So we reconstruct the complex signal by
+        // (1) slicing from the second element, (2) reversing the order,
+        // (3) taking the complex conjugate, (4) concatenating with the original
+        // input. Note that for an even input length, the last element is the
+        // Nyquist frequency which we also do not want to duplicate.
+        Eigen::DSizes<Eigen::DenseIndex, FFTRank + 1> startIndices;
+        startIndices[FFTRank] = 1;
+        auto sizes = input.dimensions();
+        if (sizes[FFTRank] % 2 == 0) {
+          sizes[FFTRank] -= 1;
+        }
+        auto cc = input.slice(startIndices, sizes).conjugate()
+          .reverse(FFTRank);
+        auto full_fft = input.concatenate(cc, FFTRank);
+
+        // Evaluate the IFFT
+        auto output = out->flat_inner_dims<float, FFTRank + 1>();
+        FFTFunctor<CPUDevice, complex64, float, Eigen::RealPart,
+                  Eigen::FFT_REVERSE, FFTRank> functor;
+        functor(ctx->eigen_device<CPUDevice>(), output, full_fft);
```

Comment:
> I was wondering whether it makes sense to stick to the functor paradigm in this case because the functors are just a templated wrapper around Eigen which are not used anywhere else.

I was thinking it would cut down on instantiations of the Eigen expression in the .o, but I just noticed that forward/backward is part of the TensorFFT template arguments, so we can't reuse the same C2C code path for forward and backward (unless we did the constant scaling for the backward outside of FFTFunctor and always did C2C forward inside of FFTFunctor). Dropping the functor is fine w/ me, but I'll defer to @benoitsteiner.

> What are your thoughts on the suggested implementation in https://github.com/tensorflow/tensorflow/pull/9029#issuecomment-292505411. And do you have any thoughts on the resulting [compiler error](http://stackoverflow.com/questions/43364086/struggling-to-evaluate-operation-in-the-tensor-module-of-eigen)? 

Argh, sorry -- not sure how I missed your comment, you were already a step ahead of me. 

Off hand without trying to compile your example, that smells like TensorFFT doesn't like having a float tensor as its input. 

Though from skimming [TensorFFT](https://bitbucket.org/eigen/eigen/src/f3a22f35b0444e03a1f65941800b9a2283de1398/unsupported/Eigen/CXX11/src/Tensor/TensorFFT.h?at=default&fileviewer=file-view-default#TensorFFT.h-30), it kind of looks like it's meant to convert `T` to `std::complex<T>` on the fly.

Have you already tried casting?

```c++
 auto op = input
  .cast<std::complex<float>>()
  .template fft<Eigen::BothParts, Eigen::FFT_FORWARD>(axes)
  .slice(startIndices, sizes);
```